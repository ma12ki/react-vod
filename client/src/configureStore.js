import { connectRoutes } from 'redux-first-router';
import { combineReducers, createStore, applyMiddleware, compose } from 'redux';
import { createEpicMiddleware } from 'redux-observable';
import createHistory from 'history/createBrowserHistory';

import { routesMap } from './routes';

import { videos } from './Videos/videos.reducers';
import { loadVideosList$ } from './Videos/videos.epics';

const history = createHistory();
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap); 

const epicMiddleware = createEpicMiddleware(loadVideosList$);

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const rootReducer = combineReducers({ location: reducer, videos });
const middlewares = applyMiddleware(middleware, epicMiddleware);
const store = createStore(rootReducer, composeEnhancers(enhancer, middlewares));

export {
    store,
};
