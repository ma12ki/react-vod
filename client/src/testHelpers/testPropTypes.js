import React from 'react';

// this has some limitations:
// https://stackoverflow.com/questions/41916992/why-does-my-jest-test-of-react-proptypes-break-when-using-multiple-unacceptable/41916993#41916993
export const testPropTypes = (component, propName, acceptableValues, unacceptableValues, otherProps) => {
    console.error = jest.fn();
    const _test = (testValues, expectError) => {
        for (let propValue of testValues) {
            console.error.mockClear();
            React.createElement(component, {...otherProps, [propName]: propValue});
            try {
                expect(console.error).toHaveBeenCalledTimes(expectError ? 1 : 0);
            } catch (err) {
                console.log(`${propName} expected to ${expectError ? '' : 'NOT'} fail for value ${propValue}`);
                throw err;
            }
        }
    };
    _test(acceptableValues || [], false);
    _test(unacceptableValues || [], true);
};
