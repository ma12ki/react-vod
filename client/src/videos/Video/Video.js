import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Link from 'redux-first-router-link';
import { Card, CardBlock, CardTitle, CardSubtitle, CardText } from 'reactstrap';

import { routesKeys } from '../routes';
import { getPayload } from '../../router.selectors';
import { getVideosEntity } from '../selectors';
import { loadVideoStart } from '../ducks';
import { Player } from '../../player';
import { FileDuration, FileSize } from '../../shared';

export class Video extends React.PureComponent {
    static propTypes = {
        video: PropTypes.shape({
            id: PropTypes.string,
            name: PropTypes.string,
            ext: PropTypes.string,
            title: PropTypes.string,
            size: PropTypes.number,
            duration: PropTypes.number,
        }),
        loadVideo: PropTypes.func.isRequired,
    }

    componentDidMount() {
        const { video } = this.props;
        if (!video.id) {
            this.props.loadVideo();
        }
    }

    render() {
        const { video } = this.props;
        const { id, name, ext, title, size = 0, duration = 0 } = video; 
        return (
            <div>
                <Card>
                    <CardBlock>
                        { id ? <Player id={id} ext={ext} /> : null }
                    </CardBlock>
                    <CardBlock>
                        <CardTitle>{name}{ext}</CardTitle>
                        <CardSubtitle>
                            <FileDuration duration={duration} />, <FileSize size={size} />
                        </CardSubtitle>
                        <CardText>{title}</CardText>
                        <Link to={{ type: routesKeys.home }}>{'<<'} Back</Link>
                    </CardBlock>
                </Card>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    const { id } = getPayload(state);
    const video = getVideosEntity(state, id);

    return {
        video,
        id,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        dispatch,
    };
};

const mergeProps = (stateProps, dispatchProps, ownProps) => {
    const { id } = stateProps;
    const { dispatch } = dispatchProps;
    return {
        ...stateProps,
        ...dispatchProps,
        ...ownProps,
        loadVideo: () => dispatch(loadVideoStart(id)),
    };
};

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(Video);
