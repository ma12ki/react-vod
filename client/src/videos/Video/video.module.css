.card.card {
    border-radius: 0;
    border-width: 0;
    background-color: transparent;
    color: #ddd;
}

.ext {
    color: #aaa;
    font-size: 14px;
}
