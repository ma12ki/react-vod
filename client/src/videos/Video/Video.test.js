import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';

import { testPropTypes } from '../../testHelpers';
import { Video } from './Video';

const mockState = {
    location: {
        routesMap: {}
    }
};
const makeWrapper = ({ video = {}, loadVideo = () => null } = {}) => shallow(<Video video={video} loadVideo={loadVideo} />);
const mountWrapper = ({ video = {}, loadVideo = () => null, mockStore = configureStore()(() => mockState) } = {}) => {
    return mount(
        <Provider store={mockStore}>
            <Video video={video} loadVideo={loadVideo} />
        </Provider>
    );
};

describe('Video', () => {

    it('matches the snapshot', () => {
        const wrapper = makeWrapper();

        expect(wrapper).toMatchSnapshot();
    });

    it('accepts "video" object', () => {
        testPropTypes(
            Video,
            'video',
            [{}, null],
            [123, 'abc'],
            { loadVideo: () => null }
        );
    });

    it('requires "loadVideo" function', () => {
        testPropTypes(
            Video,
            'loadVideo',
            [() => null],
            [123, 'abc', null, undefined],
            { video: {} }
        );
    });

    it('calls loadVideo if video ID is missing', () => {
        const loadVideo = jest.fn();
        mountWrapper({ loadVideo });

        expect(loadVideo).toHaveBeenCalled();
    });

    it('does NOT call loadVideo if video ID is present', () => {
        const video = {
            id: '123',
            ext: '.mp4'
        };
        const loadVideo = jest.fn();
        mountWrapper({ video, loadVideo });

        expect(loadVideo).not.toHaveBeenCalled();
    });

});