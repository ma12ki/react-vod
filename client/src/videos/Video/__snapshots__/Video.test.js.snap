// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video matches the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <Card
        className="card"
        tag="div"
    >
        <CardBlock
            tag="div"
        >
            <CardTitle
                tag="h4"
            >
                <span
                    className="ext"
                />
            </CardTitle>
        </CardBlock>
        <CardBlock
            tag="div"
        />
        <CardBlock
            tag="div"
        >
            <CardSubtitle
                tag="h6"
            >
                <Unknown
                    duration={0}
                />
                , 
                <Unknown
                    size={0}
                />
            </CardSubtitle>
            <CardText
                tag="p"
            />
            <Connect(Link)
                to={
                    Object {
                        "type": "vod/router/videos/HOME",
                      }
                }
            >
                &lt;&lt;
                 Back
            </Connect(Link)>
        </CardBlock>
    </Card>
</div>,
  "nodes": Array [
    <div>
      <Card
            className="card"
            tag="div"
      >
            <CardBlock
                  tag="div"
            >
                  <CardTitle
                        tag="h4"
                  >
                        <span
                              className="ext"
                        />
                  </CardTitle>
            </CardBlock>
            <CardBlock
                  tag="div"
            />
            <CardBlock
                  tag="div"
            >
                  <CardSubtitle
                        tag="h6"
                  >
                        <Unknown
                              duration={0}
                        />
                        , 
                        <Unknown
                              size={0}
                        />
                  </CardSubtitle>
                  <CardText
                        tag="p"
                  />
                  <Connect(Link)
                        to={
                              Object {
                                    "type": "vod/router/videos/HOME",
                                  }
                        }
                  >
                        &lt;&lt;
                         Back
                  </Connect(Link)>
            </CardBlock>
      </Card>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 1,
      "_context": Object {},
      "_currentElement": <Video
        loadVideo={[Function]}
        video={Object {}}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Video {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "loadVideo": [Function],
          "video": Object {},
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <Card
                    className="card"
                    tag="div"
          >
                    <CardBlock
                              tag="div"
                    >
                              <CardTitle
                                        tag="h4"
                              >
                                        <span
                                                  className="ext"
                                        />
                              </CardTitle>
                    </CardBlock>
                    <CardBlock
                              tag="div"
                    />
                    <CardBlock
                              tag="div"
                    >
                              <CardSubtitle
                                        tag="h6"
                              >
                                        <Unknown
                                                  duration={0}
                                        />
                                        , 
                                        <Unknown
                                                  size={0}
                                        />
                              </CardSubtitle>
                              <CardText
                                        tag="p"
                              />
                              <Connect(Link)
                                        to={
                                                  Object {
                                                            "type": "vod/router/videos/HOME",
                                                          }
                                        }
                              >
                                        &lt;&lt;
                                         Back
                              </Connect(Link)>
                    </CardBlock>
          </Card>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <Card
                    className="card"
                    tag="div"
          >
                    <CardBlock
                              tag="div"
                    >
                              <CardTitle
                                        tag="h4"
                              >
                                        <span
                                                  className="ext"
                                        />
                              </CardTitle>
                    </CardBlock>
                    <CardBlock
                              tag="div"
                    />
                    <CardBlock
                              tag="div"
                    >
                              <CardSubtitle
                                        tag="h6"
                              >
                                        <Unknown
                                                  duration={0}
                                        />
                                        , 
                                        <Unknown
                                                  size={0}
                                        />
                              </CardSubtitle>
                              <CardText
                                        tag="p"
                              />
                              <Connect(Link)
                                        to={
                                                  Object {
                                                            "type": "vod/router/videos/HOME",
                                                          }
                                        }
                              >
                                        &lt;&lt;
                                         Back
                              </Connect(Link)>
                    </CardBlock>
          </Card>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Video
    loadVideo={[Function]}
    video={Object {}}
/>,
}
`;
